src/Services/SubmitService/SubmitService.cs

```diff
@@ -21,28 +25,64 @@ public class SubmitService : ISubmitService
 
         private readonly SubmissionServiceConfiguration _submissionServiceConfiguration;
 
+        private readonly IDistributedCacheWrapper _distributedCache;
+
+        private readonly ISchemaFactory _schemaFactory;
+
+        private readonly IReferenceNumberProvider _referenceNumberProvider;
 
         public SubmitService(
             IGateway gateway,
             IPageHelper pageHelper,
             IWebHostEnvironment environment,
-            IOptions<SubmissionServiceConfiguration> submissionServiceConfiguration)
+            IOptions<SubmissionServiceConfiguration> submissionServiceConfiguration,
+            IDistributedCacheWrapper distributedCache,  
+            ISchemaFactory schemaFactory,
+            IReferenceNumberProvider referenceNumberProvider)
         {
             _gateway = gateway;
             _pageHelper = pageHelper;
             _environment = environment;
             _submissionServiceConfiguration = submissionServiceConfiguration.Value;
+            _distributedCache = distributedCache;
+            _schemaFactory = schemaFactory;
+            _referenceNumberProvider = referenceNumberProvider;
+        }
+
+        public async Task PreProcessSubmission(string form, string sessionGuid)
+        {
+            var baseForm = await _schemaFactory.Build(form);
+            if(baseForm.GenerateReferenceNumber)
+                _pageHelper.SaveCaseReference(sessionGuid, _referenceNumberProvider.GetReference(baseForm.ReferencePrefix), true, baseForm.GeneratedReferenceNumberMapping);
         }
 
         public async Task<string> ProcessSubmission(MappingEntity mappingEntity, string form, string sessionGuid)
         {
-            if (_submissionServiceConfiguration.FakeSubmission)
+            var baseForm = await _schemaFactory.Build(form);
+            var reference = string.Empty;
+
+            if(baseForm.GenerateReferenceNumber)
             {
+                var answers = JsonConvert.DeserializeObject<FormAnswers>(_distributedCache.GetString(sessionGuid));
+                reference = answers.CaseReference;
+            }
+
+            return _submissionServiceConfiguration.FakeSubmission ? 
+                ProcessFakeSubmission(sessionGuid, reference) :
+                await ProcessGenuineSubmission(mappingEntity, form, sessionGuid, baseForm, reference);
```

Comment:
The formatting of this should be:
condition
? positive result
: negative result;

Reply:
Done